controller/tap/server.go

```diff
@@ -493,20 +486,48 @@ func indexPodByIP(obj interface{}) ([]string, error) {
 	return []string{""}, fmt.Errorf("object is not a pod")
 }
 
-// hydrateIPMeta returns a map of expanded metadata labels for a given IP.
-// If no pod could be found for that IP, it returns an empty map.
-func (s *server) hydrateIPMeta(ip *public.IPAddress) (map[string]string, error) {
-	pod, err := s.podForIP(ip)
+// hydrateEventLabels attempts to hydrate the metadata labels for an event's
+// source and (if the event was reported by an inbound proxy) destination,
+// and adds them to the event's `SourceMeta` and `DestinationMeta` fields.
+//
+// Since errors encountered while hydrating metadata are non-fatal and result
+// only in missing labels, any errors are logged at the WARN level.
+func (s *server) hydrateEventLabels(ev *public.TapEvent) {
+	err := s.hydrateIPLabels(ev.Source.Ip, ev.SourceMeta.Labels)
 	if err != nil {
-		return nil, err
+		log.Warnf("error hydrating source labels: %s", err)
 	}
-	if pod == nil {
-		log.Debugf("no pod for IP %s", addr.PublicIPToString(ip))
-		return make(map[string]string), nil
+
+	if ev.ProxyDirection == public.TapEvent_INBOUND {
+		// Events emitted by an inbound proxies don't have destination labels,
+		// since the inbound proxy _is_ the destination, and proxies don't know
+		// their own labels.
+		err = s.hydrateIPLabels(ev.Destination.Ip, ev.DestinationMeta.Labels)
+		if err != nil {
+			log.Warnf("error hydrating destination labels: %s", err)
+		}
 	}
 
-	ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
-	return pkgK8s.GetPodLabels(ownerKind, ownerName, pod), nil
+}
+
+// hydrateIPMeta attempts to determine the metadata labels for `ip` and, if
+// successful, adds them to `labels`.
+func (s *server) hydrateIPLabels(ip *public.IPAddress, labels map[string]string) error {
+	pod, err := s.podForIP(ip)
+	switch {
+	case err != nil:
+		return err
+	case pod == nil:
+		log.Debugf("no pod for IP %s", addr.PublicIPToString(ip))
+		return nil
+	default:
+		ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
+		podLabels := pkgK8s.GetPodLabels(ownerKind, ownerName, pod)
+		for key, value := range podLabels {
+			labels[key] = value
+		}
+		return nil
+	}
```

Comment:
Sorry for the late review on this branch. I just wanted to add one style comment, but it's super minor, so no need to submit a follow-up branch. We tend to use go's [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments) as our default style guide in this code base. And using the default case of the switch statement here to handle the normal code path goes against their recommendation regarding [indent error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow). Specifically:

> Try to keep the normal code path at a minimal indentation

Since your first two cases have early returns, then you could exit the switch statement before resuming the normal code path. E.g.:

```
	pod, err := s.podForIP(ip)
	switch {
	case err != nil:
		return err
	case pod == nil:
		log.Debugf("no pod for IP %s", addr.PublicIPToString(ip))
		return nil
	}

	ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
	podLabels := pkgK8s.GetPodLabels(ownerKind, ownerName, pod)
	for key, value := range podLabels {
		labels[key] = value
	}
	return nil
```

But at that point I'd probably remove the switch statement altogether and write this as:

```
	pod, err := s.podForIP(ip)
	if err != nil {
		return err
	}

	if pod == nil {
		log.Debugf("no pod for IP %s", addr.PublicIPToString(ip))
		return nil
	}

	ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
	podLabels := pkgK8s.GetPodLabels(ownerKind, ownerName, pod)
	for key, value := range podLabels {
		labels[key] = value
	}
	return nil
```

I find that method of error handling to be easier to follow.

Reply:
Okay, thanks for the feedback --- I'll keep that in mind as a refactor if I make additional changes to this branch.